layerfetcher/layers.go

```diff
@@ -337,11 +337,16 @@ func (l *Logic) fetchLayerData(ctx context.Context, logger log.Log, layerID type
 	}
 	l.mutex.Unlock()
 
+	l.fetcher.TrackBallotsPeer(ctx, peer, ballotsToFetch)
+
 	logger.With().Debug("fetching new ballots", log.Int("to_fetch", len(ballotsToFetch)))
-	if err := l.GetBallots(ctx, ballotsToFetch); err != nil {
+	if err := l.GetBallots(ctx, ballotsToFetch, peer); err != nil {
```

Comment:
`GetBallots` calls `HandleBallotData` after response received where ballot blocks are fetched and they need peer aswell - https://github.com/spacemeshos/go-spacemesh/blob/2703-hash-not-found-new/proposals/handler.go#L399

Reply:
first of all, you don't need to pass peer to `GetBallots()` because you can get peers that has this hash from the cache in `getHahes()` and call `receiver(ctx, res.Data)` with any of them.

but you can simplify this even more by adding an interface `AddPeersFromHash()` to `system.Fetcher`. so in `HandleBallotData()` before calling `processBallot`, you can do the following
```
bidHash := b.ID().AsHash32()
h.fetcher.AddPeersFromHash(bidHash, collectHashes(b))
```
in fetcher.go for `AddPeersFromHash` you just get the peers registered for `bidHash` and add them to each of the hashes in the ballot.
the idea is, whoever peers give you this ballot ID, they should have all the hashes contained in the ballot.

`collectHashes()` is just gathering all the hashes in the ballot. this way you don't have to pass peer downstream to from `processBallot` because you have pre-registered all the hashes in the ballot.
```
func collectHashes(b *types.Ballot) []types.Hash32 {
	hashes := types.BlockIDsToHashes(ballotBlockView(b))
	if b.RefBallot != types.EmptyBallotID {
		hashes = append(hashes, b.RefBallot.AsHash32())
	}
	hashes = append(hashes, b.Votes.Base.AsHash32())
	return hashes
}
```

Reply:
thanks, done